libtextsecure/message_receiver.js

```diff
@@ -1545,24 +1546,42 @@ MessageReceiver.prototype.extend({
       window.log.info('Error getting conversation: ', envelope.source);
     }
   },
+  async handleSessionRequestMessage(envelope, content) {
+    const shouldProcessSessionRequest = await session.Protocols.SessionProtocol.shouldProcessSessionRequest(
+      envelope.source,
+      envelope.timestamp
+    );
+
+    if (shouldProcessSessionRequest) {
+      if (content.preKeyBundleMessage) {
+        await this.savePreKeyBundleMessage(
+          envelope.source,
+          content.preKeyBundleMessage
+        );
+      }
+      await session.Protocols.SessionProtocol.onSessionRequestProcessed(
+        envelope.source
+      );
+      window.log.info('sending session established to', envelope.source);
+      // We don't need to await the call below because we just want to send it off
+      window.libloki.api.sendSessionEstablishedMessage(envelope.source);
+    }
+  },
   async innerHandleContentMessage(envelope, plaintext) {
     const content = textsecure.protobuf.Content.decode(plaintext);
+    const { SESSION_REQUEST } = textsecure.protobuf.Envelope.Type;
 
-    if (content.preKeyBundleMessage) {
-      await this.savePreKeyBundleMessage(
-        envelope.source,
-        content.preKeyBundleMessage
+    if (envelope.type === SESSION_REQUEST) {
+      await this.handleSessionRequestMessage(envelope, content);
+    } else {
+      await session.Protocols.SessionProtocol.onSessionEstablished(
```

Comment:
We might want to profile this to see that we don't slow down the application when we get a lot of incoming messages.